File: src/contests/cg/puzzle/cpp/telephone-numbers.cpp
// 🎮 CodinGame Puzzle - telephone-numbers
// https://www.codingame.com/training/medium/telephone-numbers

#include <iostream>
#include <string>
#include <map>
using namespace std;

struct Node {
    map<char, Node *> ch;
};

int main() {
    int n;
    cin >> n;
    Node *root = new Node();
    long long cable = 0;
    for (int i = 0; i < n; i++) {
        string num;
        cin >> num;
        Node *node = root;
        for (char d : num) {
            if (node->ch.find(d) == node->ch.end()) {
                node->ch[d] = new Node();
                cable++;
            }
            node = node->ch[d];
        }
    }
    cout << cable << endl;
}

File: src/contests/cg/puzzle/ocaml/telephone-numbers.ml
(* 🎮 CodinGame Puzzle - telephone-numbers *)
(* https://www.codingame.com/training/medium/telephone-numbers *)

type node = { tbl : (char, node) Hashtbl.t }

let () =
  let n = int_of_string (String.trim (read_line ())) in
  let cable = ref 0 in
  let root = { tbl = Hashtbl.create 16 } in
  for _ = 1 to n do
    let number = String.trim (read_line ()) in
    let node = ref root in
    String.iter
      (fun d ->
        if not (Hashtbl.mem (!node).tbl d) then begin
          Hashtbl.add (!node).tbl d { tbl = Hashtbl.create 16 };
          incr cable
        end;
        node := Hashtbl.find (!node).tbl d)
      number
  done;
  Printf.printf "%d\n" !cable
